Autodoc

pawamoy opened the pull request:
Issue #173 

We use mkdocs to write and build the documentation, with the material theme for mkdocs, as well as the mkdocstrings plugin for auto-documentation of the source code.

I had to remove the doctoc pre-commit hook because I couldn't get its generated tables of contents to be rendered correctly in the documentation (they are a mix of HTML and Markdown). Note that this is not a great loss, as the material theme for mkdocs adds its own table of contents in the right sidebar for every page. We lose the ToC on GitHub only (and it was not really usable either because it's not sticky).

To try it, clone my branch and run `poetry run mkdocs serve`, then go to `localhost:8000`.

I could use early feedback on the following TODO points:

TODO:
- [x] Improve the code documentation: some modules have no docstrings at all, so they appear empty in the docs. The other docstrings could really be detailed a bit more. We can also benefit from the cross-reference ability of mkdocstrings, linking objects between them to improve navigation in the reference.
- [x] Customize the docs? Version 5 of mkdocs-material makes it easier to customize the docs color and icons. I'd be more inclined to let the authors/maintainers of copier do so :slightly_smiling_face: 
- [x] Split the README (aka overview in the docs) into multiple files? Create a template, generate a project, update a project, etc..., then clean it up (remove API section maybe).

![Screenshot_2020-05-06_14-49-51](https://user-images.githubusercontent.com/3999221/81178705-ec4d7b80-8fa8-11ea-9ed9-5cfc4a435eab.png)

Yajo commented on README.md:
```diff
@@ -1,4 +1,5 @@
-## Think this library is awesome? Vote with a 👍 to include it in the awesome-python list: https://github.com/vinta/awesome-python/pull/1350
+**Think this library is awesome? Vote with a 👍 to include it in the awesome-python
+list: https://github.com/vinta/awesome-python/pull/1350**
```
```suggestion
list: https://github.com/DoronCohen/awesome-python/pull/1**
```

That repo is dead according to the issue itself.

Yajo commented on .gitignore:
```diff
@@ -33,6 +33,7 @@ pip-wheel-metadata
 # Documentation
 */_build/*
 */site/*
+site/
```
Possibly you meant this:

```suggestion
/site
```

Isn't this covered by line 35?

pawamoy replied:
I was surprised too, but no, line 35 does not ignore the root `site` directory :confused:

pawamoy commented:
> I'd appreciate some example(s) on how that linking works.

Sure. Simply put, you can link to any heading anywhere else in the site with the classic Markdown syntax, and since each Python object is rendered with a heading, you can link to them:
- with a custom link title: <code>[\`EnvOps\` class][copier.config.objects.EnvOps]</code>
- or directly: `[copier.config.objects.EnvOps][]`.

This will work both in mkdocs Markdown pages and in Python docstrings (as long as they are valid Markdown).

And "any heading" means you can also link to other headings, for example: [See the Get Started section in Contributing][get-started]`.

This will continue to work even if headings are moved from one page to another, or if a Python object is moved from one module to another. mkdocstrings will warn if it finds references that cannot be resolved, for example `[see this][this-heading-does-not-exist]`.

> typing annotations appears nowhere in the docs

They do not appear in the generated docs signatures, no, and it's not configurable in mkdocstrings (yet). However we're supposed to write the docstrings in the Google format. When doing that, mkdocstrings render the docstring with tables for parameters, return value and exceptions, using type annotations in the signature, something like this: https://pawamoy.github.io/mkdocstrings/reference/handlers/python/#mkdocstrings.handlers.python.PythonRenderer.render

It means that if we follow the Google-style and document each parameter in the docstrings, they will indeed be rendered in the generated docs :slightly_smiling_face: 

> Yes, now with proper docs, README should be much shorter.

OK, I'll try to split the README then.

> There's one thing missing: please add some Github action workflow that uploads these docs. AFAICS it should be quite easy to do: https://www.mkdocs.org/user-guide/deploying-your-docs/

Good idea!

> If there were a way to produce some browsable artifacts for PRs and upload docs just for pushes in the master branch, that'd be just awesome.

Just to be sure, you want to deploy the docs for each push on master? I'd favor deploying when pushing tags only, and it's related to the next point ->

> BTW, any chances to get docs per version (from now on)? 

Unfortunately, I didn't find any mature solution allowing to version the docs. About the previous point, if we have only one deployed version at a time, I think it would be best to document the latest release, and not the current commit (master branch). Of course if we manage to deploy multiple versions, having docs for each release **and** for the master branch is the best.

Now, for the things I tried:
- [mike](https://github.com/jimporter/mike), but I don't remember what the result looks like. I'll try it again.

Also, I don't quite agree with their idea:

> mike is built around the idea that once you've generated your docs for a particular version, you should never need to touch that version again.

When vulnerabilities are found, you *absolutely need* to modify the docs for the impacted versions, to warn users. Same when a version is deprecated and not maintained, you *really should* display a banner discouraging people to use this version, and to upgrade to a newer one. And again, in the development version (master branch), you *should* warn users that this version is the development one, and this warning needs to go away once it's released. Finally, mistakes can be made while writing docs, and I think developers should always be able to correct them, even if this version of the docs has already been deployed. Wrong docs are worse than no docs. Anyway, that's just my opinion :smile: 

- [mkdocs-versioning](https://github.com/zayd62/mkdocs-versioning), but the project seems to be on hold, and I don't quite like how it works: it writes each version in a subfolder (:+1:) so they can be reached using sub-URLs (:+1:), but to switch from one version to another you have to click on a `Versions` link in the sidebar, then click again on the desired version (:-1:). You cannot switch version while staying on the same page or anchor (:-1:).

Yajo commented:
> When vulnerabilities are found, you _absolutely need_ to modify the docs for the impacted versions, to warn users. Same when a version is deprecated and not maintained, you _really should_ display a banner discouraging people to use this version, and to upgrade to a newer one. And again, in the development version (master branch), you _should_ warn users that this version is the development one, and this warning needs to go away once it's released.

I think your concerns about mike are more theoretical that practical. :thinking: 

See, if we rebuild the gh site on each commit to master, AFAIK we have to rebuild the whole site. That includes static pages for previous versions.

Also, those problems you mention can be easily fixed by maintaining pages like "supported versions" or "vulnerabilities" statically in the latest version only (if that's possible).

If we build on each commit, and the build includes all tags, then if that commit has a tag and we are able to default to that latest tag, it should be just OK.

Also I can see that mkdocs itself is documented in readthedocs and, if you browse https://readthedocs.org/projects/mkdocs/ in several versions, you'll see it has a floating version switcher. Maybe that feature is added automatically by readthedocs? If so, we could use it for docs and forget about the problem.

Otherwise we can leave this improvement for the future when https://github.com/mkdocs/mkdocs/issues/193 is fixed upstream.

> They do not appear in the generated docs signatures, no, and it's not configurable in mkdocstrings (yet).

I opened https://github.com/mkdocs/mkdocs/issues/2097 upstream, for now using the google style should be enough, although double-writing types is not fun. Maybe we can just use Google syle without types, and just benefit from that upstream improvement if/when implemented. Anyways, your IDE should give you enough type hints when needed.

pawamoy commented:
Oh, @Yajo the auto-documentation is solely the responsibility of mkdocstrings, not mkdocs. You should have opened an issue there :slightly_smiling_face:

Yajo commented:
Oops thanks! I'm new to all that stuff...

pawamoy commented:
> I think your concerns about mike are more theoretical that practical.

They are indeed :smile: I need to play a bit with it again.

> if we rebuild the gh site on each commit to master, AFAIK we have to rebuild the whole site. That includes static pages for previous versions.

Not sure this is possible with mike:
> mike is designed to produce one version of your docs at a time

...and each command it provides expect one and only one specific version as argument.

> Also, those problems you mention can be easily fixed by maintaining pages like "supported versions" or "vulnerabilities" statically in the latest version only (if that's possible).

Not a lot of people would go to the "vulnerabilities" page to check it, even less think that is such a page to check.

---

Anyway, I forgot about ReadTheDocs, but yes, this could work just fine, and we would have versions support!

Yajo commented:
> each command it provides expect one and only one specific version as argument.

Can we do something like this? 

```python
for v in git("tag"):
    mike(v)
```

> Anyway, I forgot about ReadTheDocs, but yes, this could work just fine, and we would have versions support!

What would you prefer then? You are more expert here in documenting. 😁

Yajo commented:
Hey friend, I've been gradudated by @pykong to owner here (thanks pal! :heart:). I have all needed permissions now to link with readthedocs if needed, although I think you're the expert in that sense. Think you could continue here?

pawamoy commented:
Hey friend! I was glad to hear good news from @pykong and new :slightly_smiling_face:!

Yes, I'll eventually continue the work on this PR, unfortunaltely my time is quite limited these days because of work :confused:

Yajo commented:
OK, don't worry, we're not in a hurry ;)

El mié., 17 jun. 2020 a las 21:37, Timothée Mazzucotelli (<
notifications@github.com>) escribió:

> Hey friend! I was glad to hear good news from @pykong
> <https://github.com/pykong> and new 🙂!
>
> Yes, I'll eventually continue the work on this PR, unfortunately my time
> is quite limited these days because of work 😕
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/copier-org/copier/pull/203#issuecomment-645610930>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAHNXDNI3XDMZC7XPV4ZPTTRXESPVANCNFSM4M2MFCPQ>
> .
>


-- 
Jairo Llopis

Yajo commented:
Hi friend, this is a kind reminder that whenever you can continue with this, it would be awesome. :blush:  Copier is getting some traction and I see some questions like #231 that could easily fit into some FAQ in the docs, if we had them.

Thanks for your help! :heart:

pawamoy commented:
Hi, thanks for the reminder! I'll dedicate some time this evening!

pawamoy commented:
@Yajo, I think you can now already enable the project on ReadTheDocs. Did you do such a thing already? Otherwise I can guide you. You just have to login with GitHub on ReadTheDocs, then "Import" the project. Make sure to check the "Advanced" checkbox during the setup, and select MkDocs instead of Sphinx.

pawamoy commented:
Alright, I split the README and created three pages: creating a template, generating a project, and updating a project. The rest stayed in the README, which is also the overview in the docs. I got rid of the auto-generated table of contents which were not rendered correctly in the docs. I also added docstrings and type annotations to the code. It's not perfect but it's a good start!

Waiting for RTD to be linked to the project, and for your review of course :slightly_smiling_face:

Yajo commented on copier/cli.py:
```diff
@@ -172,12 +222,13 @@ class CopierUpdateSubApp(cli.Application):
     """
 
-    only_diff = cli.Flag(
+    only_diff: cli.Flag = cli.Flag(
```
This type seems to not be detected properly:

![2020-07-24_08-13](https://user-images.githubusercontent.com/973709/88365219-9c02e000-cd85-11ea-8136-3d40f32739ff.png)

pawamoy replied:
Indeed, my intuition is that the `@CopierApp.subcommand("update")` decorator has some responsibility in this :thinking:

Yajo replied:
It's not something critical, but maybe we should open an issue upstream... Do you feel like doing that?

pawamoy replied:
Sure, issue opened here: https://github.com/pawamoy/pytkdocs/issues/60!

Yajo commented on mkdocs.yml:
```diff
@@ -0,0 +1,47 @@
+site_name: copier
+site_description: Library and command-line utility for rendering projects templates.
+site_url: https://pykong.github.io/copier
+repo_url: https://github.com/pykong/copier
+repo_name: pykong/copier
```
```suggestion
site_url: https://copier.readthedocs.io/
repo_url: https://github.com/copier-org/copier
repo_name: copier-org/copier
```

Yajo commented on CHANGELOG.md:
```diff
@@ -3,30 +3,6 @@
-<!-- prettier-ignore-end -->
-</details>
-
 ### Version 4.0.0 (2020-06)
```
I just realized I forgot to update the latest changelogs.

```suggestion
### Version 4.0.2 (2020-07-21)

[All changes here](https://github.com/copier-org/copier/milestone/11?closed=1). Summary:

- Fix wrong templated default answers classification, which produced some questions being ignored.

### Version 4.0.1 (2020-06-23)

[All changes here](https://github.com/copier-org/copier/milestone/10?closed=1). Summary:

- Fix wrong prompt regression when updating.
- Remove redundant `dst` fixture in tests.

### Version 4.0.0 (2020-06)
```

Yajo commented on copier/cli.py:
```diff
@@ -25,6 +33,22 @@ def _wrapper(*args, **kwargs):
 
 
 class CopierApp(cli.Application):
+    """
+    The Plumbum CLI application.
```
I tried always to follow [PEP 257](https://www.python.org/dev/peps/pep-0257/) which says:

>  The summary line may be on the same line as the opening quotes or on the next line. The entire docstring is indented the same as the quotes at its first line (see example below).

So, this should be:

```suggestion
    """The Plumbum CLI application.
```

I noticed it doesn't alter the mkdocs result, so please change this.

pawamoy replied:
Based on the PEP description, I follow it as well :stuck_out_tongue: 

But of course, let's keep the existing style! I was sleepy when I committed last night ^^

Yajo commented on copier/cli.py:
```diff
@@ -42,21 +66,21 @@ class CopierApp(cli.Application):
     data: AnyByStrDict = {}
 
-    answers_file = cli.SwitchAttr(
+    answers_file: cli.SwitchAttr = cli.SwitchAttr(
```
I'm surprised about these changes. I'd say this is a bug on mkdocs because mypy doesn't yell about this. `cli.SwitchAttr` is a class, so it's pretty obvious that this attribute will have that class as a type.

Any ideas on this subject?

pawamoy replied:
Well, `mkdocstrings` is not as powerful as `mypy`, and is only using actual type annotations, not deducing them from the value of the attribute. So if you want the type to appear in the docs, you *have* to annotate it.

pawamoy replied:
Actually `pytkdocs` could be improved to use the attribute's type if no annotation is found. I'll open the related feature request 🙂

Yajo commented on copier/config/factory.py:
```diff
@@ -65,7 +67,8 @@ def make_config(
     **kwargs,
 ) -> ConfigData:
-    """Provides the configuration object, merged from the different sources.
+    """
+    Provides the configuration object, merged from the different sources.
```
According to https://www.python.org/dev/peps/pep-0257/#multi-line-docstrings this was OK:

```suggestion
    """Provides the configuration object, merged from the different sources.
```

Yajo commented on copier/main.py:
```diff
@@ -151,7 +152,12 @@ def copy(
 
 
 def copy_local(conf: ConfigData) -> None:
+    """
+    Use the given configuration to generate a project.
```
```suggestion
    """Use the given configuration to generate a project.
```

Yajo commented on copier/main.py:
```diff
@@ -199,7 +205,13 @@ def copy_local(conf: ConfigData) -> None:
 
-def update_diff(conf: ConfigData):
+def update_diff(conf: ConfigData) -> None:
+    """
+    Use the given configuration to update a project.
```
```suggestion
    """Use the given configuration to update a project.
```

Yajo commented on copier/main.py:
```diff
@@ -273,6 +285,20 @@ def get_source_paths(
     render: Renderer,
     must_filter: Callable[[StrOrPath], bool],
 ) -> List[Tuple[Path, Path]]:
+    """
+    Get the paths to all the files to render.
```
```suggestion
    """Get the paths to all the files to render.
```

Yajo commented on copier/main.py:
```diff
@@ -294,6 +320,15 @@ def get_source_paths(
 
 
 def render_folder(rel_folder: Path, conf: ConfigData) -> None:
+    """
+    Render a complete folder.
```
```suggestion
    """Render a complete folder.
```

Yajo commented on copier/main.py:
```diff
@@ -319,7 +354,16 @@ def render_file(
     must_skip: CheckPathFunc,
 ) -> None:
-    """Process or copy a file of the skeleton."""
+    """
+    Process or copy a file of the skeleton.
```
```suggestion
    """Process or copy a file of the skeleton.
```

Yajo commented on copier/main.py:
```diff
@@ -346,12 +390,35 @@ def render_file(
 
 
 def files_are_identical(src_path: Path, dst_path: Path, content: Optional[str]) -> bool:
+    """
+    Tell whether two files are identical.
```
```suggestion
    """Tell whether two files are identical.
```

Yajo commented on copier/main.py:
```diff
@@ -346,12 +390,35 @@ def render_file(
 
 
 def overwrite_file(conf: ConfigData, dst_path: Path, rel_path: Path) -> bool:
+    """
+    Handle the case when there's an update conflict.
```
```suggestion
    """Handle the case when there's an update conflict.
```

Yajo commented on copier/main.py:
```diff
@@ -361,6 +428,14 @@ def overwrite_file(conf: ConfigData, dst_path: Path, rel_path: Path) -> bool:
 
 
 def run_tasks(conf: ConfigData, render: Renderer, tasks: Sequence[Dict]) -> None:
+    """
+    Run the given tasks.
```
```suggestion
    """Run the given tasks.
```

Yajo commented on copier/tools.py:
```diff
@@ -180,7 +184,8 @@ def match(path: StrOrPath) -> bool:
 
 def get_migration_tasks(conf: ConfigData, stage: str) -> List[Dict]:
-    """Get migration objects that match current version spec.
+    """
+    Get migration objects that match current version spec.
```
```suggestion
    """Get migration objects that match current version spec.
```

Yajo commented:
Could you please rebase and/or squash? It seems like the bots got stuck, I don't know why? :thinking:

pawamoy commented:
Done!
